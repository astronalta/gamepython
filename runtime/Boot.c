/*
 * Copyright (c) 2010 Matt Fichman
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to
 * deal in the Software without restriction, including without limitation the
 * rights to use, copy, modify, merge, publish, distribute, sublicense, and/or
 * sell copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 * 
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 * 
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, APEXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
 * FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS
 * IN THE SOFTWARE.
 */

#include "String.h"
#include <stdio.h>
#include <stdlib.h>

void boot_print_str(String* string) {
    // Write string to stdout.  This is function is here for convenience's 
    // sake.  Once a full-fledged IO framework has been written, this function
    // will really only be useful for simple output.

    fwrite(string->data, 1, string->length, stdout);
    fflush(stdout);
}

void boot_print_int(Int integer) {
    // Print an integer to stdout.  This function is here only to run initial
    // tests on the compiler, and isn't part of the public API.

#ifdef DARWIN
    fprintf(stdout, "%lld", integer);
#else
    fprintf(stdout, "%ld", integer);    
#endif
    fflush(stdout);
}

void boot_print_char(Char character) {
    // Print a character to stdout.  This function is not part of the public 
    // API for the Apollo library.

    fputc(character, stdout);
    fflush(stdout);
}

void boot_dummy(int a, int b, int c) {
    boot_print_int(a);
    boot_print_int(b);
    boot_print_int(c);
}
