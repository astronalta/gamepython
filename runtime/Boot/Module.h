/*
 * Copyright (c) 2010 Matt Fichman
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to
 * deal in the Software without restriction, including without limitation the
 * rights to use, copy, modify, merge, publish, distribute, sublicense, and/or
 * sell copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 * 
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 * 
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, APEXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
 * FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS
 * IN THE SOFTWARE.
 */

#ifndef JOGO_BOOT_H
#define JOGO_BOOT_H

#include "Primitives.h"

Ptr Boot_malloc(Int size);
Ptr Boot_calloc(Int size);
void Boot_mzero(Ptr val, Int size);
void Boot_memcpy(Ptr src, Ptr dst, Int size);
void Boot_free(Ptr memory);
void Boot_abort();

#endif
