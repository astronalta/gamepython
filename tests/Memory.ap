import Boot


Counter < Object {

    num private Int

    @init(number Int) {
        num = number
        boot_print_str('alloc: ')
        boot_print_int(num)
        boot_print_str('\n')
    }

    @destroy() {
        boot_print_str('free: ')
        boot_print_int(num)
        boot_print_str('\n')
    }
}

StrongReference < Object {
    obj Empty
    num Int

    @init(param Empty, number Int) {
        num = number
        boot_print_str('alloc: ')
        boot_print_int(num)
        boot_print_str('\n')
    }

    @destroy() {
        boot_print_str('free: ')
        boot_print_int(num)
        boot_print_str('\n')
    }
}

WeakReference < Object {
    obj weak Empty
    num Int

    @init(number Int) {
        num = number
        boot_print_str('alloc: ')
        boot_print_int(num)
        boot_print_str('\n')
    }

    @destroy() {
        boot_print_str('free: ')
        boot_print_int(num)
        boot_print_str('\n')
    }
}

test1() Counter {
    ret Counter(3)
}

test2(obj Counter) Counter {
    ret obj
}

test3(a Counter, b Counter) {
}

main() Int {
    obj3 = test1()

    let obj1 = Counter(1) {
        let obj2 = Counter(2) {
            
        }
    }

    obj4 = test2(Counter(4))
    obj5 Empty = Counter(5)
    test3(Counter(6), Counter(7)) 
    test3(Counter(8), Counter(9)) 
    if Counter(10) {
        Counter(11)
    }

    while not Counter(12) {
    }

    let obj6 = StrongReference(Counter(13), 14) {
        obj5 = Counter(15)
    }

    obj7 = WeakReference(16)
    obj8 = StrongReference(obj7, 17)
    obj7.obj = obj8

    ret 0
}
