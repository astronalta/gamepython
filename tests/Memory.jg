# jgtest exec on

import Boot

Counter < Object {

    num private Int

    @init(number Int) {
        num = number
        Boot::print_str('alloc: ')
        Boot::print_int(num)
        Boot::print_str('\n')
    }

    @destroy() {
        Boot::print_str('free: ')
        Boot::print_int(num)
        Boot::print_str('\n')
    }
}

StrongReference < Object {
    obj Empty
    num Int

    @init(param Empty, number Int) {
        num = number
        Boot::print_str('alloc: ')
        Boot::print_int(num)
        Boot::print_str('\n')
    }

    @destroy() {
        Boot::print_str('free: ')
        Boot::print_int(num)
        Boot::print_str('\n')
    }
}

WeakReference < Object {
    obj mutable weak Empty
    num Int

    @init(number Int) {
        num = number
        Boot::print_str('alloc: ')
        Boot::print_int(num)
        Boot::print_str('\n')
    }

    @destroy() {
        Boot::print_str('free: ')
        Boot::print_int(num)
        Boot::print_str('\n')
    }
}

test1() Counter {
    ret Counter(3)
}

test2(obj Counter) Counter {
    ret obj
}

test3(a Counter, b Counter) {
}

main() Int {
    obj1 = Counter(0)
    obj3 = test1()

    let obj1 = Counter(1) {
        let obj2 = Counter(2) {
        }
    }

    obj4 = test2(Counter(4))
    obj5 mutable Empty = Counter(5)
    test3(Counter(6), Counter(7)) 
    test3(Counter(8), Counter(9)) 
    if Counter(10) {
        Counter(11)
    }

    while not Counter(12) {
    }

    let obj6 = StrongReference(Counter(13), 14) {
        obj5 = Counter(15)
    }

    obj7 = WeakReference(16)
    obj8 = StrongReference(obj7, 17)
    obj7.obj = obj8

    ret 0
}
