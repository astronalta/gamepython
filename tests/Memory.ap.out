alloc: 3
alloc: 1
alloc: 2
free: 2
free: 1
alloc: 4
alloc: 5
alloc: 6
alloc: 7
free: 6
free: 7
alloc: 8
alloc: 9
free: 8
free: 9
alloc: 10
free: 10
alloc: 11
free: 11
alloc: 12
free: 12
alloc: 13
alloc: 14
free: 13
alloc: 15
free: 5
free: 14
alloc: 16
alloc: 17
free: 17
free: 16
free: 15
free: 4
free: 3
